src/NLog/Layouts/Layout.cs

```diff
@@ -282,13 +295,26 @@ internal void Initialize(LoggingConfiguration configuration)
                 // layout is thread agnostic if it is thread-agnostic and 
                 // all its nested objects are thread-agnostic.
                 this.threadAgnostic = true;
+                this.stackTraceUsage = StackTraceUsage.None;
+
                 foreach (object item in ObjectGraphScanner.FindReachableObjects<object>(this))
                 {
-                    if (!item.GetType().IsDefined(typeof(ThreadAgnosticAttribute), true))
+                    if (this.threadAgnostic)
                     {
-                        this.threadAgnostic = false;
-                        break;
+                        if (!item.GetType().IsDefined(typeof(ThreadAgnosticAttribute), true))
+                        {
+                            this.threadAgnostic = false;
```

Comment:
I see we lost the `break` here. AFAIK `IsDefined(typeof(ThreadAgnosticAttribute),` not very cheap. 

Maybe we should loop two times (same list) over `ObjectGraphScanner.FindReachableObjects<object>(this)`, and has `breaks` in both loops?